sky/dag.py

```diff
@@ -2,74 +2,253 @@
+            ValueError: If the task already exists in the DAG or if its name
+            is already used.
+        """
+        if task in self.tasks:
+            raise ValueError(f'Task {task.name} already exists in the DAG.')
         self.graph.add_node(task)
         self.tasks.append(task)
+        if task.name is not None:
+            if task.name in self._task_name_lookup:
+                raise ValueError(
+                    f'Task name "{task.name}" is already used in the DAG.')
+            self._task_name_lookup[task.name] = task
+
+    def remove(self, task: Union['task.Task', str]) -> None:
+        """Remove a task from the DAG.
+
+        Args:
+            task: The Task object or name of the task to remove.
 
-    def remove(self, task: 'task.Task') -> None:
+        Raises:
+            ValueError: If the task is still being depended on by other tasks.
+        """
+        task = self._get_task(task)
+
+        dependents = list(self.graph.successors(task))
+        for dependent in dependents:
+            self.remove_dependency(dependent, task)
+
+        # TODO(andy): Stuck by optimizer's wrong way to remove dummy sources
+        # and sink nodes. And we need a proper way to always show task's name
+        # if dependents:
+        #     dependent_names = ', '.join(
+        #         [dep.name for dep in dependents if dep.name is not None])
+        #     raise ValueError(f'Task {task.name} is still being depended on '
+        #                      f'by tasks {dependent_names!r}. Try to remove
+        #                      'the dependencies first.')
+
+        self.dependencies.pop(task, None)
         self.tasks.remove(task)
         self.graph.remove_node(task)
+        if task.name is not None:
+            self._task_name_lookup.pop(task.name, None)
+
+    def add_dependency(self, dependent: TaskOrName,
+                       dependency: TaskOrName) -> None:
+        """Add a single dependency for a task.
+
+        This method adds a new dependency without removing existing ones.
+
+        Args:
+            dependent: The task that depends on another.
+            dependency: The task that the dependent task depends on.
+
+        Raises:
+            ValueError: If a task tries to depend on itself or if the
+            dependency task is not in the DAG.
+        """
+        dependent = self._get_task(dependent)
+        dependency = self._get_task(dependency)
+
+        if dependent == dependency:
+            raise ValueError(f'Task {dependent.name} cannot depend on itself.')
+        if dependency not in self.tasks:
+            raise ValueError(
+                f'Dependency task {dependency.name} is not in the DAG.')
+
+        self.graph.add_edge(dependency, dependent)
+        if dependent not in self.dependencies:
+            self.dependencies[dependent] = set()
+        self.dependencies[dependent].add(dependency)
 
     def add_edge(self, op1: 'task.Task', op2: 'task.Task') -> None:
```

Comment:
Lets remove this function? seems like it is identical to `add_dependency`

Reply:
To maintain back compatibility?

Reply:
Good point! How about rename our api to `add_edge`, `remove_edge`, `set_edge` instead

Reply:
I think that would be inconsistent with the dependencies on the yaml side. Wouldn't a better way be to mark `add_edge` as deprecated and then remove it in the next release?

Reply:
That sounds great. For deprecate comments, pls reference to: https://github.com/skypilot-org/skypilot/blob/a4e2fcd438d70373377c85bcbec1b185ef04c99f/sky/provision/gcp/instance.py#L636-L637